### PR TITLE
generate launching images based on ES themes

### PR DESCRIPTION
**Pull request by meleu**
Here is the scriptmodule we talked about [in the forum](https://retropie.org.uk/forum/topic/8029/where-s-the-retropie-setup-api-docs/3).

### CODE REVIEW

**Comment by joolswills** [comment]
thanks. Few general comments before a full review - there are a lot of multi-line statements - although it's useful to keep lines short, I think perhaps it's a little too much in some cases here. Also please use 4 spaces for indentation (even for multiline).

eg

```
    dialog \
        --backtitle "$__backtitle" \
        --no-items \
        --menu "$text" \
        22 86 16 \
        $@ \
        2>&1 >/dev/tty
```

but in this case I don't think it needs splitting - and it's enough to just do it as follow - note that "$@" should be quoted.

```
    dialog --backtitle "$__backtitle" --no-items --menu "$text" 22 86 16 "$@"  2>&1 >/dev/tty
```

when you have an array like

```
    options=(
        'something' \
        'something' \
    )
```

you can leave out the `\` characters.

**Comment by joolswills** [comment]
also function names all need to end with the module name - so they are guaranteed unique.

**Comment by meleu** [comment]
OK. I'll try to do as you said right now.

**Comment by meleu** [comment]
From [line 300](https://github.com/meleu/RetroPie-Setup/blob/31991e9b5a9cac6d464834ea0aea458c62dcb15d/scriptmodules/supplementary/launchingimages.sh/#L300) to [360](https://github.com/meleu/RetroPie-Setup/blob/31991e9b5a9cac6d464834ea0aea458c62dcb15d/scriptmodules/supplementary/launchingimages.sh/#L360) the code has a deep indentation. I would like to use nested if-elif (rather than case) to avoid 4 spaces of ugliness. Are you OK with it?

**Comment by joolswills** [comment]
I would probably avoid it completely, by separating things out a bit into modular functions. I would use a case statement though still.

**Comment by joolswills** [comment]
BTW check out the scraper and _load_config_scraper function - for how we load settings there .

the multiple settings loading / saving is probably overkill - a single saved config should be enough - but up to you - but `loadModuleConfig` should be useful here and allow you to split out the code a bit more.

**Comment by meleu** [comment]
Yeah, I tried to think in a modular function logic to avoid it, but I need access to several local variables that isn't accessible globally (I'm assuming that a scriptmodule can't declare global variables).
I'm working on the style first...
P.S.: OK. I'll look at the scraper function.

**Comment by joolswills** [comment]
That's why you can use loadModuleConfig and have a function that does that to load the variables so you can split it out a bit (means a bit more reloading, but it's not noticable). note with local variables also in bash that functions below the function they are defined in can read them

**Comment by joolswills** [comment]
in the case of the scraper, we reload them in each function, as from commandline each function can be called independently - eg to automate and so on.

**Comment by joolswills** [comment]
Looking at your code I would split out these

```
1 "Image generation settings"
2 "Generate launching images"
3 "View slideshow of all current launching images"
4 "View a specific system's launching image"
```

into separate functions that can be called independently (so they would just reload the current config as needed).

**Comment by meleu** [comment]
I will take a look at this loadModuleConfig stuff. Thanks for the tips!

**Comment by meleu** [comment]
Sorry if I'm making this conversation a bash class, but I think I misunderstood what you said here:

> note with local variables also in bash that functions below the function they are defined in can read them

To check it I've made this test:
```
[prompt]$ cat test.sh
#!/bin/bash

function func_a() {
    local var_a="this is declared in func_a()"
    echo "func_a(): var_a=$var_a"
}

function func_b() {
    echo "func_b(): var_a=$var_a"
}

func_a
func_b
echo "no functions: var_a=$var_a"

[prompt]$ ./test.sh
func_a(): var_a=this is declared in func_a()
func_b(): var_a=
no functions: var_a=
```

Maybe you meant variables declared in a function but **not** declared explicitly as `local`, right? If yes, is it an acceptable practice for scriptmodules?

(I'm studying the `loadModuleConfig` and wanna use it, I'm asking it just to learn more and more. :) )

**Comment by joolswills** [comment]
Your code is as expected. When I said below I mean when one function is called from another.

```
#!/bin/bash

function func_a() {
    local var_a="this is declared in func_a()"
    echo "func_a(): var_a=$var_a"
    func_b
}

function func_b() {
    echo "func_b(): var_a=$var_a"
}

func_a
echo "no functions: var_a=$var_a"
```

**Comment by meleu** [comment]
Oh, I've got it. Thanks for the clarification.

I'm working on the scriptmodule right now, and I'm making available the multiple config files because there are some known good configs for different themes.

I'm thinking in putting some optimized config files for some themes in the [generate-launching-images repository](https://github.com/meleu/generate-launching-images), to make things easier to the user.

**Comment by meleu** [comment]
I think it's ready for your review. Thanks for all the tips.
Cheers!

File: scriptmodules/supplementary/launchingimages.sh [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+    getDepends "${depends[@]}"
+}
+
+function sources_launchingimages() {
+    gitPullOrClone "$md_inst" "https://github.com/meleu/generate-launching-images.git"
+}
+
+function install_launchingimages() {
+    cd "$md_inst"
+    chmod a+x generate-launching-images.sh
```
Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: as this is a single script, you could probably just have an install_bin function

Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: the file permissions could also be committed to git, which would remove the need for the chmod

File: scriptmodules/supplementary/launchingimages.sh [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+}
+
+function _show_images_launchingimages() {
+    [[ -f "$1" ]] || return 1
+
+    local image="$1"
+    local timeout=5
+    local is_list=0
+
+    [[ "$(file -i $1)" =~ text/plain ]] && is_list=1
```
Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: This seems odd - why do you need to check the type of file ? could be unreliable - can you not just use a parameter if it is to process a list 

no need to do a regular expression for a match like this either - can use *text/plan*

File: scriptmodules/supplementary/launchingimages.sh (hunk 1 above)
Comment (meleu) [inline scriptmodules/supplementary/launchingimages.sh]: good catch! Should I commit these changes or will you do?

File: scriptmodules/supplementary/launchingimages.sh [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+            --autozoom \
+            $([[ "$is_list" -eq 1 ]] && echo --list) \
+            "$image" </dev/tty &>/dev/null
+    fi
+}
+
+function _dialog_menu_launchingimages() {
+    local text="$1"
+    shift
+    [[ -z "$@" ]] && return 1
```
Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: You use "$#" below - would be a better test here also.

File: scriptmodules/supplementary/launchingimages.sh [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+    fi
+
+    dialog --backtitle "$__backtitle" --no-items --menu "$text" 22 86 16 "$@" 2>&1 >/dev/tty
+}
+
+function _set_theme_launchingimages() {
+    _dialog_menu_launchingimages \
+        "List of available themes" \
+        $("$md_inst/generate-launching-images.sh" --list-themes) \
+    || echo "\$theme"
```
Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: I would have thought the last line should be indented too, if splitting onto lots of lines.

Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: Not a big deal, just doesn't really match any other code style in retropie

File: scriptmodules/supplementary/launchingimages.sh [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+        choice=$("${cmd[@]}" "${options[@]}" 2>&1 >/dev/tty)
+
+        [[ -z "$choice" ]] && break
+
+        if [[ "$choice" = "config_file" ]]; then
+            config_file=$(_manage_config_file_launchingimages)
+            continue
+        fi
+
+        eval iniSet "$choice" \"$(_set_${choice}_launchingimages)\"
```
Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: why eval ?

Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: oh. you are returning a variable name in a called function - wouldn't it be best if the functions just echoed the variable value - this seems more logical.

Comment (meleu) [inline scriptmodules/supplementary/launchingimages.sh]: it wasn't an iniSet before. I'm changing...

File: scriptmodules/supplementary/launchingimages.sh [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+            )"
+            extension ".$extension"
+            loading_text "\"$loading_text\""
+            press_button_text "\"$press_button_text\""
+            loading_text_color "$loading_text_color"
+            press_button_text_color "$press_button_text_color"
+            show_timeout "$( [[ -n "$no_ask" ]] && echo "don't show (see no_ask)" || echo "$show_timeout seconds")"
+            no_ask "$( [[ -n "$no_ask" ]] && echo true || echo false)"
+            no_logo "$( [[ -n "$no_logo" ]] && echo true || echo false)"
+            logo_belt "$( [[ -n "$logo_belt" ]] && echo true || echo false)"
```
Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: I'm not sure I like this for readability over how we build options lists in other modules, and requiring commandline tools like echo etc for each option. You don't need to change it, just mentioning.

File: scriptmodules/supplementary/launchingimages.sh [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+    local config_file
+
+    iniConfig ' = ' '"' "$md_inst/.current.cfg"
+
+    while true; do
+        eval $(_load_config_launchingimages)
+
+        options=( 
+            config_file "$(
+               [[ "$config_file" =~ $theme\.cfg$ ]] && echo "$(basename "$config_file")"
```
Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: can you avoid a re here ? eg something like

`[[ "$config_file" == *"$theme.cfg" ]] && echo "$(basename "$config_file")"`

**Comment by joolswills** [comment]
Thanks for your work on this - I will give it a test and merge when you have finished with the changes. Please can you squash the changes when you are ready. Cheers.

File: scriptmodules/supplementary/launchingimages.sh (hunk 4 above)
Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: I see the logic in this now - you are indenting the parameters for the function, but not the `|| echo` - so feel free to leave this as is. It's just due to this module having a lot more line splitting making it it look quite different from other retropie code.

File: scriptmodules/supplementary/launchingimages.sh (hunk 2 above)
Comment (meleu) [inline scriptmodules/supplementary/launchingimages.sh]: Done: https://github.com/RetroPie/RetroPie-Setup/pull/1881/commits/73e07f28483f075a50580bda2857eb702291dd8f

File: scriptmodules/supplementary/launchingimages.sh (hunk 5 above)
Comment (meleu) [inline scriptmodules/supplementary/launchingimages.sh]: Done: https://github.com/RetroPie/RetroPie-Setup/pull/1881/commits/fd3ce817a0fb7b92f983095245fe01b4ec1daae6

File: scriptmodules/supplementary/launchingimages.sh (hunk 3 above)
Comment (meleu) [inline scriptmodules/supplementary/launchingimages.sh]: Done: https://github.com/RetroPie/RetroPie-Setup/pull/1881/commits/67b3bf54360f3ec7cc089c3478f46c10dac881d9

File: scriptmodules/supplementary/launchingimages.sh (hunk 4 above)
Comment (meleu) [inline scriptmodules/supplementary/launchingimages.sh]: Done: https://github.com/RetroPie/RetroPie-Setup/pull/1881/commits/f757c39ca030d3846e511e61568d8c869b8134f3

File: scriptmodules/supplementary/launchingimages.sh (hunk 7 above)
Comment (meleu) [inline scriptmodules/supplementary/launchingimages.sh]: Done: https://github.com/RetroPie/RetroPie-Setup/pull/1881/commits/d935c6c8fd0e5df3b14ab0b26c4cf3c7c32c2ae9

File: scriptmodules/supplementary/launchingimages.sh (hunk 5 above)
Comment (meleu) [inline scriptmodules/supplementary/launchingimages.sh]: Fixing: https://github.com/RetroPie/RetroPie-Setup/pull/1881/commits/07a695735429f6c2b83a91f5fd19e729249c8db1

Thanks to your tip here: https://github.com/RetroPie/RetroPie-Setup/pull/1881#issuecomment-279080194
Thanks!

File: scriptmodules/supplementary/launchingimages.sh (hunk 1 above)
Comment (meleu) [inline scriptmodules/supplementary/launchingimages.sh]: I'm planning to add some optimized `.cfg` files for some themes.
Examples:
- material theme: use `--solid-bg-color` and `--press-button-text-color white`.
- clean-look theme: use `--logo-belt`

Is the `install_bin` approach good for this case also?

Comment (meleu) [inline scriptmodules/supplementary/launchingimages.sh]: Done: https://github.com/RetroPie/RetroPie-Setup/pull/1881/commits/cdac6bbf949041668cb9ca71a280b43a1db4eeb4

**Comment by meleu** [comment]
I think it's ready for your tests.

File: scriptmodules/supplementary/launchingimages.sh [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+rp_module_section="exp"
+rp_module_flags="noinstclean"
+
+function depends_launchingimages() {
+    local depends=(imagemagick librsvg2-bin)
+    if isPlatform "x11"; then
+        getDepends feh
+    else
+        getDepends fbi
+    fi
```
Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: these should be depends+=(feh) otherwise apt will be called twice.

**Comment by joolswills** [comment]
Spotted the above when testing - please can you squash the changes.

**Comment by joolswills** [comment]
the system selection logic if different from other dialogs in retropie. cancel reverts it back to "all". I htink all should be the first option, and cancel reverts it back to what it was before.

**Comment by joolswills** [comment]
if choosing option 4 - when no images are present, it does nothing - perhaps a dialog ?

When only one theme is installed, it says "list of available themes: carbon" and has a yes / no which seems odd - why not just a menu selection with a single item ?

If choosing generate and no theme was selected it shows an error. perhaps this should be done at the "image generation settings" stage ?

the image generation settings "cancel" button should be renamed I think - since it doesn't actually cancel anything that has been entered  perhaps "return" or "exit"

**Comment by joolswills** [comment]
I see in the code there is some reason for the yes no dialog instead of a menu - looks unneeded - just provide an empty string as the item ?

You could also just have a numbered menu which is what we do in other dialogs.

**Comment by meleu** [comment]
I'm quite busy currently. But I'll do the changes you pointed as soon as possible! ;-)

**Comment by joolswills** [comment]
No rush :-) Apart from the points I mentioned above it all looks good, and works very well - we should be good to go once that's sorted. Thanks.

**Comment by meleu** [comment]
@joolswills I found some time to do what you suggest.

> if choosing option 4 - when no images are present, it does nothing - perhaps a dialog ?

Done.


> When only one theme is installed, it says "list of available themes: carbon" and has a yes / no which seems odd - why not just a menu selection with a single item ?

Done. The problem is that the Raspbian's dialog version is 1.2-20140911 and I wrote the scriptmodule in Linux Mint 18 (dialog version 1.3-20160209). The dialog version on Raspbian has a small bug when using `--menu` option with `--no-items` and only one tag.

I need the `--no-items` option because I wanna get exactly what dialog puts in the stdout, (this is simpler than getting the tag number and then get the corresponding item).

> If choosing generate and no theme was selected it shows an error. perhaps this should be done at the "image generation settings" stage ?

I did this way thinking in a curious user navigating through the dialogs. If the user enter in the "image generation settings" and then cancel without a theme selected (to view the current images or just exit the script), everything is fine, there's no problem with this action. The really problem happens if the user chose to create the images with no theme selected.

If it's not a big problem, I prefer to leave this way...


> the image generation settings "cancel" button should be renamed I think - since it doesn't actually cancel anything that has been entered perhaps "return" or "exit"

I changed to "Back".


> the system selection logic is different from other dialogs in retropie. cancel reverts it back to "all". I think all should be the first option, and cancel reverts it back to what it was before.

Done.

**Comment by meleu** [comment]
Oh! I forgot to mention about the `librsvg2-bin` dependency. It is needed to convert the logo `.svg` files.

- in my RetroPie on Linux Mint 18 it **is** a dependency.
- when testing in my 4.1 on Raspi installation, it's unnecessary (imagemagick is enough).
- when testing in a 4.0 on Raspi of a friend of mine, `librsvg2-bin` was a dependency and installed a bunch of X/Wayland libraries (can't remember exactly, but it was 80+ MB of packages). Maybe an `apt-get update/upgrade` would avoid it, but I don't know the best way to do it in a scriptmodule.

Could you please sort this last thing?
(As if the normal problems weren't enough, my youngest son has pneumonia...)

**Comment by joolswills** [comment]
>Done. The problem is that the Raspbian's dialog version is 1.2-20140911 and I wrote the scriptmodule in Linux Mint 18 (dialog version 1.3-20160209). The dialog version on Raspbian has a small bug when using --menu option with --no-items and only one tag.

I would do as I mentioned before and not use no-items - just use numbered items. You can just get the value from the array. We don't use no-items anywhere else, so it is more fitting with the retropie menu styling.

If you are busy I can make this change for you.

regarding the dependencies - it should be the same for retropie 4 and 4.1 - if testing installing manually for testing - make sure you do `apt-get install --no-install-recommends` or else more will be installed than is needed. getDepends does this automatically.

Sorry to hear about your son - best wishes for a quick recovery!

**Comment by meleu** [comment]
OK. Now the dialog menus have numbers.
A touch of algebra helped to make it simple. :-)

Regarding the `librsvg2-bin` depedency, I'm pretty sure I don't have it installed on my raspi/RetroPie 4.1 and I can convert the `.svg` logo files. So it's **not** a dependency in this platform.
On the other hand, if this package isn't installed on a x86 RetroPie, the convert fails when working with `.svg` files.

Summing up: I think the scriptmodule is complete now. Phew!

File: scriptmodules/supplementary/launchingimages.sh [hunk 9]
```diff
@@ -68,16 +67,21 @@ function _show_images_launchingimages() {
...
-        dialog --backtitle "$__backtitle" --no-items --menu "$text" 22 86 16 "${options[@]}" 2>&1 >/dev/tty
-    fi
+    i=1
+    for opt in "$@"; do
+        options+=( "$i" "$opt" )
+        ((i++))
+    done
+    choice=$(dialog --backtitle "$__backtitle" --menu "$text" 22 86 16 "${options[@]}" 2>&1 >/dev/tty) || return
+    i=$[ 2 * choice - 1 ]
+    echo "${options[$i]}"
```
Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: can just do `echo "${options[choice*2-1]}"`

Comment (meleu) [inline scriptmodules/supplementary/launchingimages.sh]: done

Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: btw please use (not needed in this case but for reference) `$(( ... ))` for arithmetic expansion

Comment (joolswills) [inline scriptmodules/supplementary/launchingimages.sh]: to clarify - `$[ ... ]` is obsolete (from old bash I think).

**Comment by joolswills** [comment]
Thanks. I squashed via github interface and merged.
